<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Thierry Gosselin" />

<meta name="date" content="2019-11-06" />

<title>RADseq analysis cloud tutorial</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><div><img src="logo.png"></img>RADseq cloud tutorial</div></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="radseq_cloud_tutorial.html">Tutorial</a>
</li>
<li>
  <a href="https://github.com/thierrygosselin">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">RADseq analysis cloud tutorial</h1>
<h4 class="author">Thierry Gosselin</h4>
<h4 class="date">2019-11-06</h4>

</div>


<p><strong>Outline:</strong></p>
<ul>
<li><a href="#intro">Tutorial introduction</a></li>
<li><a href="#hardware">Hardware requirements</a></li>
<li><a href="#setup">Personal computer setup</a></li>
<li><a href="#ec2config">Amazon EC2 configurations</a></li>
<li><a href="#cloudstore">Storing data in the cloud</a></li>
<li><a href="#spot">Starting a spot instance</a></li>
<li><a href="#connect">Connection with the instance</a></li>
<li><a href="#stacks">Run Stacks</a></li>
<li><a href="#userstudio">Run RStudio</a></li>
<li><a href="#ami">Creating your own AMI</a></li>
<li><a href="#RADseq_software">Install GBS/RADseq software</a></li>
<li><a href="#rstudio">Install R and Rstudio Server</a></li>
<li><a href="#usefullink">Useful</a></li>
</ul>
<div id="intro" class="section level1">
<h1>Tutorial introduction</h1>
<p>This step-by-step tutorial is a risk-free paragliding tandem flight into the cloud. We will guide you through the process of leveraging the cloud to analyse your RADseq data.</p>
<p>For an evolutionary biologist, restriction site-associated DNA sequencing (RADseq) and new genomic tools are great opportunities, but the fast-changing field can be very challenging to follow. The most important tool for RADseq, after your brain and the sequencer, is your computer. RADseq analysis can be very demanding on personal computers, there can be not enough CPU and/or memory and the analysis can crash or worse, your computer is now solely dedicated to this task and nothing else. If you’re lucky enough to have access to a large up-to-date university computer clusters this might be a good solution but they come with their <em>share</em> of problems.</p>
<p>We chose Amazon Elastic Compute Cloud (EC2) because: <em>(i) it’s relatively easy</em>, their’s extensive documentation and big community to ask for help; <em>(ii) it’s relatively cheap</em> to use machines with lots of CPU and memory. Using the cloud offers 2 more advantages:</p>
<ol style="list-style-type: decimal">
<li>you can write off computer and software cost as an <em>operating expense</em> rather than a capital or service expenditure (depending on countries and universities).</li>
<li>you control everything: no need to ask IT to update one of the numerous dependenices you’ll likely required…</li>
</ol>
<p><strong>Assumptions</strong></p>
<p>Although you can go through this tutorial with <em>copy and paste</em> commands, knowing very basic <em>Terminal</em> commands will be useful:</p>
<ul>
<li><a href="http://practicalcomputing.org/files/PCfB_Appendices.pdf">Reference card</a>: quite useful to get through the jargon and main commands.</li>
<li><a href="https://thoughtbot.com/blog/the-magic-behind-configure-make-make-install">Good reading</a>: to understand the basic steps to install software with the terminal.</li>
<li>Online courses: e.g. with <a href="https://www.datacamp.com">DataCamp</a>.</li>
</ul>
<p><strong>Recommendations</strong>:</p>
<p>In the beginning most students will feel lost and very intimidated by everything related to computers, software and RADseq analysis. Commonly the end result is that your scientific work flow is impeded. If you’re at the Master and PhD level and all the RADseq stuff is overwhelming and you’re not sleeping at night, consider:</p>
<ul>
<li><p>you should focus first on biology, second, on bioinformatics (a computer is the main tool you’ll likely use in your future) and if you have time, the wet-lab part. Now, don’t get me wrong, I think the wet-lab is the most important step, but unless you want to become a wet-lab technician, my advice is stay away as much as humanly possible from the wet-lab and leave it to experts.</p></li>
<li><p>asking for help:</p>
<ul>
<li>ask your director to be supervised by a PostDoc who knows the ropes in exchange for authorship on the paper(s).</li>
<li>ask for outside help, there’s a lot of sequencing facilities that now offers to do the extractions, librairies, sequencing and genotyping.</li>
</ul></li>
</ul>
<p>Thierry Gosselin</p>
</div>
<div id="hardware" class="section level1">
<h1>Hardware requirements</h1>
<div id="desktop-computer" class="section level2">
<h2>Desktop computer</h2>
<p>Used for fine tuning the workflow, running preliminary analysis with small data sets and making ready-to-publish figures. Here are some specs that will help make the RADseq pipeline run smoothly:</p>
<ul>
<li>Recommended OS architecture: <a href="http://en.wikipedia.org/wiki/UNIX">UNIX</a>, (e.g. Apple Computers or PCs running Linux)</li>
<li><a href="http://en.wikipedia.org/wiki/Central_processing_unit">CPU</a>: &gt; 8 cores architecture</li>
<li><a href="http://en.wikipedia.org/wiki/Memory_(computers)">Memory</a>: minimum of 16 GB RAM</li>
<li>Disk: 500 GB to 1 TB <a href="http://en.wikipedia.org/wiki/Solid-state_drive">SSD drive</a></li>
<li>External <a href="http://en.wikipedia.org/wiki/Hard_disk_drive">HDD drive</a> for storing analyses and backup (you need Terabytes not Megabytes).</li>
</ul>
</div>
<div id="cloud-computer" class="section level2">
<h2>Cloud computer</h2>
<p><strong>Understanding Amazon Elastic Cloud Compute (EC2) Instances</strong></p>
<p>Amazon EC2 instances comes in different <a href="http://en.wikipedia.org/wiki/Central_processing_unit">CPU</a> and <a href="http://en.wikipedia.org/wiki/RAM_memory">memory</a> configurations (<a href="http://aws.amazon.com/ec2/instance-types/">instances types</a> and <a href="http://aws.amazon.com/ec2/pricing/">prices</a>). For molecular biologists 3 purchasing options will be interesting: spot, on-demand and reserved instances.</p>
<div id="spot-instances" class="section level3">
<h3><a href="http://aws.amazon.com/ec2/purchasing-options/spot-instances/">Spot Instances</a></h3>
<p>Spot Instances allow you to name your own price for Amazon EC2 computing capacity. Yes, you read correctly, no need to see the web site or look in the documentation. You simply bid on unused Amazon EC2 instances and run your instances for as long as your bid remains higher than the current Spot Price. You specify the maximum hourly price that you are willing to pay to run a particular instance type. The Spot Price fluctuates based on supply and demand for instances, but customers will never pay more than the maximum price they have specified. If the Spot Price moves higher than a customer’s maximum price, the customer’s instance will be shut down by Amazon EC2 (<a href="http://youtu.be/wcPNnUo60pc">how to manage interruption</a>). Other than those differences, <strong>Spot Instances perform exactly the same as On-Demand or Reserved Instances.</strong> This pricing model provides the most cost-effective option for obtaining compute capacity with interruption-tolerant tasks.</p>
</div>
<div id="on-demand-instances" class="section level3">
<h3><a href="https://aws.amazon.com/ec2/purchasing-options/">On-Demand Instances</a></h3>
<p>On-Demand Instances let you pay the specified hourly rate for the instances you use with no long-term commitments or upfront payments. This type of purchasing option is recommended for applications that cannot be interrupted.</p>
</div>
<div id="reserved-instances" class="section level3">
<h3><a href="http://aws.amazon.com/ec2/purchasing-options/reserved-instances/">Reserved Instances</a></h3>
<p>Reserved Instances let you make a low, one-time, upfront payment for an instance, reserve it for a one or three year term, and pay a significantly lower hourly rate for that instance. For applications that have steady state needs, Reserved Instances can provide savings of up to 70% compared to using On-Demand Instances. This is probably the least interesting option for biologists.</p>
</div>
<div id="alternatives-to-amazon-ec2" class="section level3">
<h3>Alternatives to Amazon EC2</h3>
<p>Do you think Google Cloud might be a better solution for you? <a href="https://cloud.google.com/products/compute-engine/">See their compute engine here</a>.</p>
</div>
<div id="example-of-instance" class="section level3">
<h3>Example of instance</h3>
<p>Here is one instance that I recommend for best price/performance ratio -&gt; <a href="http://aws.amazon.com/ec2/instance-types/">Amazon EC2 instance i3.8xlarge</a></p>
<ul>
<li>CPU: 32 <a href="http://aws.amazon.com/ec2/faqs/#What_is_an_EC2_Compute_Unit_and_why_did_you_introduce_it">ECU</a> (99 vCPU) cores architecture</li>
<li>Memory: 244 GB RAM</li>
<li>Disk: 4x1900 GB of SSD (you can also add extra 600-1000GB <a href="http://aws.amazon.com/ebs/pricing/">EBS volume</a>)</li>
<li>Storage: <a href="http://aws.amazon.com/s3/">Amazon S3</a> for short term backup and to transfer next-gen data to instance. <a href="https://aws.amazon.com/glacier/">Amazon Glacier</a> for off-site and long term storage.</li>
<li><strong>On demand instance:</strong> 2,50 $US per hour</li>
<li><strong>Spot instance:</strong> 0,9 $US per hour!</li>
</ul>
</div>
</div>
</div>
<div id="setup" class="section level1">
<h1>Personal computer setup</h1>
<p>Linux and macOS can follow guidelines in <a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html">this tutorial</a> to install GBS/RADseq software on your personal computer. The codes below will help you get the required amazon tools to access AMAZON cloud services from your computer.</p>
<div id="bashprofile" class="section level2">
<h2>Start up script</h2>
<p><strong>The shell start up script and PATH to programs</strong></p>
<p>To make things a little easier to talk to your computer, each time you open the <em>Terminal</em> a <a href="http://en.wikipedia.org/wiki/Bash_(Unix_shell)">shell start up scripts</a> tells your computer where to look for programs. The path for your programs can be modified in your shell start up script. When your computer is searching for programs, it looks into these <a href="http://en.wikipedia.org/wiki/PATH_(variable)">path</a>:</p>
<pre class="bash"><code># In your Terminal
$PATH</code></pre>
<p>The output vary depending on computer OS and version. Sometimes, it will also say: <code>No such file or directory</code> (no worries, see below).</p>
<p>Use the <code>pwd</code> (print working directory) command to know exactly where you are!</p>
<p>The name of the shell startup file differs across platforms. It’s usually called <code>.bash_profile</code>. Filenames beginning with <code>.</code> are reserved for the system and are <strong>invisible/hidden</strong>. Not all text editors are configured to see those files by default.</p>
<p>Find your shell start up script with the following command:</p>
<pre class="bash"><code># In your Terminal
ls -al ~ | grep profile</code></pre>
<p>If this returns nothing (blank), you don’t have a shell start up script. Create one with this command</p>
<pre class="bash"><code>sudo touch $HOME/.bash_profile
# $HOME points to your home directory</code></pre>
<p>To modify, you can use <a href="http://www.barebones.com/products/bbedit/index.html">BBEdit</a> to open or make and modify hidden items (using the option <em>Show hidden items</em> on the open file screen). With Linux, use Vi! With most unix system, <code>sudo nano</code> will get the job done.</p>
<p>After modifying your shell start up script to reload it, always run the command:</p>
<pre class="bash"><code># Terminal
source $HOME/.bash_profile</code></pre>
</div>
<div id="install-amazon-command-line-tools" class="section level2">
<h2>Install Amazon command line tools</h2>
<p><a href="https://docs.aws.amazon.com/en_pv/cli/latest/userguide/install-macos.html">Universal Command Line Interface for Amazon Web Services</a></p>
<ul>
<li><a href="https://docs.aws.amazon.com/en_pv/cli/latest/userguide/install-windows">Windows instructions</a></li>
<li><a href="https://docs.aws.amazon.com/en_pv/cli/latest/userguide/install-linux.html">Linux instructions</a></li>
<li><a href="https://docs.aws.amazon.com/en_pv/cli/latest/userguide/install-macos">macOS instructions</a>. Here is the short version:</li>
</ul>
<pre class="bash"><code>cd ~/Downloads #the bundled installer doesn&#39;t support installing to paths that contain spaces
curl &quot;https://s3.amazonaws.com/aws-cli/awscli-bundle.zip&quot; -o &quot;awscli-bundle.zip&quot;
unzip awscli-bundle.zip
sudo ./awscli-bundle/install -i /usr/local/aws -b /usr/local/bin/aws</code></pre>
</div>
<div id="sign-up" class="section level2">
<h2>Sign up</h2>
<p>Few more steps before using Amazon Web Services (AWS)</p>
<ol style="list-style-type: decimal">
<li><p>Go to <a href="http://aws.amazon.com" class="uri">http://aws.amazon.com</a> and click <strong>Sign In to the Console</strong>.</p></li>
<li><p>Follow the on-screen instructions.</p></li>
</ol>
<p>Note: If you have used Amazon services before (e.g. to buy books) use the same username and password, the process will be fast!</p>
</div>
<div id="get-your-keys" class="section level2">
<h2>Get your keys</h2>
<p>To drive Amazon’s computers you need 2 keys: an <strong>access</strong> and a <strong>secret</strong> keys <a href="https://docs.aws.amazon.com/en_pv/general/latest/gr/aws-security-credentials.html">Instructions on how to get your Keys</a>.</p>
<p>Your 2 keys are stored in the <a href="https://console.aws.amazon.com/iam/home?#security_credential">security credentials section</a> under your name in the upper right corner on the amazon console. Although you can retrieve your access key ID from the Your Security Credentials page, you can’t retrieve your Secret Access Key. Therefore, if you can’t find your Secret Access Key, you’ll need to create a new one before using CLI tools.</p>
<p>If you don’t want to specify your access keys every time you issue a command, using the <code>--aws-access-key</code> an <code>--aws-secret-key</code> (or <code>-O</code> and <code>-W</code>) options, you have 2 options:</p>
<ul>
<li>store your 2 keys using the following environment variables in your shell startup script (the <code>.bash_profile</code> file <a href="#bashprofile">discussed above</a>):</li>
</ul>
<pre class="bash"><code># copy/paste the 2 lines:
export AWS_ACCESS_KEY=your-aws-access-key-id 
export AWS_SECRET_KEY=your-aws-secret-key</code></pre>
<ul>
<li>configure directly by following the instructions in the Terminal:</li>
</ul>
<pre class="bash"><code>aws configure</code></pre>
<p>These will be necessary:</p>
<ul>
<li><strong>AWS Access Key ID:</strong> generated above</li>
<li><strong>AWS Secret Access Key:</strong> generated above</li>
<li><strong>region:</strong> us-east-1 (<a href="https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html">for other regions</a>)</li>
<li><strong>Default output format:</strong> text</li>
</ul>
</div>
<div id="start-the-engine" class="section level2">
<h2>Start the engine !</h2>
<p>Make sure everything is properly configured and that your computer now talks <em>Amazon language</em>:</p>
<pre class="bash"><code>aws ec2 describe-regions # this will output different regions available</code></pre>
</div>
<div id="problems" class="section level2">
<h2>Problems ?</h2>
<p>To help you solve computer related problems and start focusing back on biology:</p>
<ol style="list-style-type: decimal">
<li>Read the book: <a href="http://practicalcomputing.org">Practical Computing for Biologists</a> (go for paper, because the digital edition from Sinauer is useless).</li>
<li><a href="https://www.biostars.org">Biostar</a></li>
<li><a href="http://seqanswers.com">SEQanswers and SEQanswers wiki</a></li>
<li>You can always ask <a href="https://www.google.com">Google</a> for help.</li>
<li>Take some online classes (e.g. with <a href="https://www.datacamp.com">DataCamp</a>)</li>
<li><a href="https://www.datacamp.com/community/tutorials/aws-ec2-beginner-tutorial">DataCamp as a good tutorials for Amazon EC2</a></li>
</ol>
</div>
</div>
<div id="ec2config" class="section level1">
<h1>Amazon EC2 configurations</h1>
<p><strong>Steps before requesting/starting an Instance</strong></p>
<p>I’m sparing your sanity with most of the details, if you really want the information click on the link for the documentation.</p>
<ul>
<li><a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/ec2-key-pairs.html">Key pairs</a>: necessary to authenticate communications between Amazon and your computer. It’s similar to pairing a Bluetooth speaker with your phone.</li>
<li><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-vpc.html">VPC</a></li>
<li><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-subnet.html">Subnet</a></li>
<li><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-internet-gateway.html">Internet gateways</a></li>
<li><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-route.html">Route tables</a></li>
<li><a href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html">Security group</a>: required to control who’s talking to your instance and computer.</li>
</ul>
<div id="key-pairs" class="section level2">
<h2>Key pairs</h2>
<p>The code below will:</p>
<ul>
<li>Create a <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/ec2-key-pairs.html">key pairs</a> named <code>radseq_keys</code> in the working directory</li>
<li>Include a restriction on the key</li>
</ul>
<pre class="bash"><code>aws ec2 create-key-pair --key-name radseq_keys --query &#39;KeyMaterial&#39; --output text &gt; radseq_keys.pem
sudo chmod 400 radseq_keys.pem # computer password required</code></pre>
<p>Get the description of your keypairs with <code>aws ec2 describe-key-pairs</code></p>
</div>
<div id="create-a-vpc" class="section level2">
<h2>Create a VPC</h2>
<pre class="bash"><code>aws ec2 create-vpc --cidr-block 10.0.0.0/16</code></pre>
<p>Write down the vpc id that starts with <code>vcp-</code>:</p>
<pre class="bash"><code>vpc=&quot;vpc-0493d12477c2c2b51&quot;</code></pre>
<p>Modify the VPC to enable DNS hostnames:</p>
<pre class="bash"><code>aws ec2 modify-vpc-attribute --vpc-id $vpc --enable-dns-support
aws ec2 modify-vpc-attribute --vpc-id $vpc --enable-dns-hostnames</code></pre>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-vpc.html">VPC Documentation</a></p>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/modify-vpc-attribute.html">modify-vpc-attribute documentation</a></p>
</div>
<div id="create-a-subnet" class="section level2">
<h2>Create a subnet</h2>
<pre class="bash"><code>aws ec2 create-subnet --availability-zone us-east-1a --cidr-block 10.0.0.0/16 --vpc-id $vpc</code></pre>
<p>Write down the subnet id starting with <code>subnet-</code>:</p>
<pre class="bash"><code>subnet=&quot;subnet-0dee7de0007fac906&quot;</code></pre>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-subnet.html">subnet documentation</a></p>
</div>
<div id="create-an-internet-gateway" class="section level2">
<h2>Create an internet gateway</h2>
<pre class="bash"><code>aws ec2 create-internet-gateway</code></pre>
<p>Write down the gateway starting with <code>igw-</code>:</p>
<pre class="bash"><code>igw=&quot;igw-06f188634be4cb4ab&quot;</code></pre>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-internet-gateway.html">Internet gateways documentation</a></p>
<p>Attach the internet gateway to the VPC</p>
<pre class="bash"><code>aws ec2 attach-internet-gateway --internet-gateway-id $igw --vpc-id $vpc</code></pre>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/attach-internet-gateway.html">attach-internet-gateway documentation</a></p>
</div>
<div id="modify-route-tables" class="section level2">
<h2>Modify Route Tables</h2>
<p>Get the route tables id:</p>
<pre class="bash"><code>aws ec2 describe-route-tables</code></pre>
<p>Write down the route table starting with <code>rtb</code>:</p>
<pre class="bash"><code>rtb=&quot;rtb-0b8297dd315ded4ff&quot;</code></pre>
<p>Modify the route tables to allow the connection</p>
<pre class="bash"><code>aws ec2 create-route --route-table-id $rtb --destination-cidr-block 0.0.0.0/0 --gateway-id $igw</code></pre>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/create-route.html">create-route function documentation</a></p>
</div>
<div id="security-group" class="section level2">
<h2>Security group</h2>
<p><a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/using-network-security.html">Security group doc</a></p>
<p><strong>1. Create a security group named <em>radseq</em>:</strong></p>
<pre class="bash"><code>aws ec2 create-security-group --description &quot;radseq analysis&quot; --group-name radseq --vpc-id $vpc</code></pre>
<p>Write down the <strong>security group ID</strong>, it starts with <code>sg-</code>:</p>
<pre class="bash"><code>sg=&quot;sg-0c6ffc9bb018f9005&quot;</code></pre>
<ul>
<li>Delete a security group: <code>aws ec2 delete-security-group</code>.</li>
<li>Describe Security Group: <code>aws ec2 describe-security-groups</code>.</li>
</ul>
<p><strong>2. Communication with the instances</strong></p>
<p>Communication with your instance through <strong>port 22</strong> needs to be open. This is reserved for Secure Shell (<a href="http://en.wikipedia.org/wiki/Secure_Shell">SSH</a>) and if you’re planning on using RStudio, <strong>port 8787</strong> also needs to be open.</p>
<p>Communication security depends if your planning on using your instance from one computer/devices or more.</p>
<ul>
<li><p><strong>Use your external IP address:</strong> e.g. If you plan on using the instance from your office computer only, associate your external IP address for increase security. Several ways to get your external IP address:</p>
<ul>
<li><p>In the Terminal, use this code: <code>curl -s checkip.dyndns.org|sed -e 's/.*Current IP Address: //' -e 's/&lt;.*$//'</code>.</p></li>
<li><p>Google search <code>my ip address</code> (the first result displayed will be your external, or public, IP address).</p></li>
<li><p><a href="https://checkip.amazonaws.com">amazon tool to get your IP address</a></p></li>
</ul></li>
</ul>
<p>Add <code>/32</code> prefix <strong>after</strong> your IP address:</p>
<pre class="bash"><code># for port 22
aws ec2 authorize-security-group-ingress --group-id $sg --protocol tcp --port 22 --cidr 203.0.113.0/32 # enable SSH from your IP address.

# for port 8787
aws ec2 authorize-security-group-ingress --group-id $sg --protocol tcp --port 8787 --cidr 203.0.113.0/32 # enable SSH from your IP address.</code></pre>
<ul>
<li><strong>Use <code>0.0.0.0/0</code> if your planing to use the instance from different computers</strong></li>
</ul>
<p>e.g. you plan on using another computer or phone to check the state of the computations, use this command instead:</p>
<pre class="bash"><code># for port 22
aws ec2 authorize-security-group-ingress --group-id $sg --protocol tcp --port 22 --cidr 0.0.0.0/0

# for port 8787
aws ec2 authorize-security-group-ingress --group-id $sg --protocol tcp --port 8787 --cidr 0.0.0.0/0</code></pre>
<p><strong>To describe the security group:</strong></p>
<pre class="bash"><code>aws ec2 describe-security-groups --group-ids $sg</code></pre>
</div>
<div id="more-tutorials" class="section level2">
<h2>More tutorials</h2>
<ul>
<li><a href="http://aws.amazon.com/documentation/">All Amazon Web Services documentation</a></li>
<li><a href="http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/Welcome.html">Command Line Reference</a></li>
<li><a href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EC2_GetStarted.html">Amazon Elastic Compute Cloud (EC2)</a></li>
<li><a href="http://awsdocs.s3.amazonaws.com/EC2/latest/ec2-qrc.pdf">Quick Reference Card</a></li>
<li><a href="http://aws.amazon.com/training/intro_series/">Instructional Videos and links to free lab to practice</a></li>
</ul>
</div>
</div>
<div id="cloudstore" class="section level1">
<h1>Storing data in the cloud</h1>
<p><strong>Uploading your RADseq data in the cloud</strong></p>
<p>After trying to download your Illumina lanes to your computer or university computer clusters, you finally understand the meaning of <em>omic</em> jargon: <strong>avalanche</strong>, <strong>deluge</strong> and <strong>tsunami</strong>.</p>
<p>Moving biological data around is a challenge and it isn’t going away, technologies in the -omic fields are constantly evolving and producing more data… Accessibility, Expendability, Redundancy and Reliability: this is where <a href="http://aws.amazon.com/s3/">Amazon Simple Storage Service (S3)</a> come in to play.</p>
<p><img src="figures/s3_bucket_flowchart.png" width="872" style="display: block; margin: auto;" /></p>
<p><strong>Use Amazon S3 as:</strong></p>
<ul>
<li>a middle man for moving data around (e.g. computer &lt;-&gt; s3 &lt;-&gt; Amazon EC2 instances)</li>
<li>a cheap, out-of-site and in the cloud backup solution (more on this later)</li>
<li>well you’re already using Amazon S3 if you have Dropbox or Netflix services… these services are nice GUI front-end of Amazon S3!</li>
</ul>
<p><strong>Notes:</strong></p>
<ul>
<li><strong>Vocabulary specific for Amazon S3:</strong> In kitchens, you have cabinets with drawers filled with kitchen tools. In computers, you have directories with folders filled with files. With Amazon S3 you have <strong>buckets</strong> filled with <strong>objects</strong>. The object can be any files: music, photos or .fastq files!</li>
<li>You’re not charged for creating buckets, only for the content you put in the buckets. Do you want to keep originals and use Amazon S3 as a backup? Use the <em>Reduced Redundancy Storage (RRS)</em> feature to reduce the cost by 20% storage on Amazon S3.</li>
<li>Data transfer IN Amazon S3 : FREE</li>
<li>Data transfer OUT Amazon S3 TO Amazon EC2 in your region: FREE</li>
<li>Data transfer OUT Amazon S3 TO your computer: see <a href="http://aws.amazon.com/s3/">documentation</a> and <a href="http://aws.amazon.com/s3/pricing/">pricing</a>.</li>
<li>For RADseq data archiving, see <a href="https://aws.amazon.com/glacier/">Amazon Glacier</a>.</li>
</ul>
<div id="upload-data-with-your-browser" class="section level3">
<h3>Upload data with your browser</h3>
<ol style="list-style-type: decimal">
<li><a href="http://docs.aws.amazon.com/AmazonS3/latest/gsg/GetStartedWithS3.html">Start here</a></li>
<li>Manage your S3 buckets/objects directly from your <a href="https://console.aws.amazon.com/s3">browser</a>.</li>
</ol>
<p><strong>Get the most out of Amazon S3 with the free applications:</strong></p>
<p><a href="http://cyberduck.io">Cyberduck</a>:</p>
<ul>
<li>To view your Amazon S3 bucket content.</li>
<li>Transfer files.</li>
<li>Synchronize transfer between your computer and your bucket.</li>
<li><a href="https://trac.cyberduck.io/wiki/help/en/howto/s3">Cyberduck-Amazon S3 how to</a>.</li>
<li><a href="https://trac.cyberduck.io/raw-attachment/wiki/help/en/Cyberduck%20Quick%20Reference.pdf">Cyberduck’s Quick Refence Guide</a>.</li>
</ul>
<p>Linux users will find <a href="https://filezilla-project.org">FileZilla</a> very useful.</p>
</div>
<div id="upload-data-using-the-terminal" class="section level3">
<h3>Upload data using the Terminal</h3>
<p>The Terminal can be very powerful and useful for uploading large files to your new Amazon S3 bucket. The Amazon tools installed on you computer earlier can take advantage of <a href="https://docs.aws.amazon.com/en_pv/AmazonS3/latest/dev/mpuoverview.html">multipart upload</a> of Amazon.</p>
<p><strong>Further Amazon S3 readings:</strong></p>
<ul>
<li><a href="http://awsdocs.s3.amazonaws.com/S3/latest/s3-qrc.pdf">Quick Reference Card</a></li>
<li><a href="http://aws.amazon.com/documentation/s3/">Amazon S3 Documentation</a></li>
<li><a href="http://www.molecularecologist.com/2013/08/how-to-backup-and-store-your-next-generation-sequencing-ngs-data/">How to backup and store NGS data</a></li>
<li><a href="http://aws.amazon.com/s3/faqs/">FAQ</a></li>
</ul>
</div>
</div>
<div id="spot" class="section level1">
<h1>Starting a spot instance</h1>
<p>Below, we show how to start a spot instance using the terminal (the command line). You could also do it with the console, but it’s much faster with the command line. Before requesting a Spot Instance, we will get the <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/using-spot-instances-history.html">Pricing History</a> of the instance over the last month to help make a price decision.</p>
<div id="pricing-history-console" class="section level2">
<h2>Pricing History (console)</h2>
<p>This is copy/paste of <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/using-spot-instances-history.html">Amazon doc:</a></p>
<ol style="list-style-type: decimal">
<li>Open the Amazon EC2 console at <a href="https://console.aws.amazon.com/ec2/" class="uri">https://console.aws.amazon.com/ec2/</a>.</li>
<li>On the navigation pane, choose <strong>Spot Requests</strong>.</li>
<li>If you are new to Spot Instances, you see a welcome page. Choose <strong>Get started</strong>, scroll to the bottom of the screen, and then choose <strong>Cancel</strong>.</li>
<li>Choose <strong>Pricing History</strong>. By default, the page displays a graph of the data for Linux t1.micro instances in all Availability Zones over the past day. Move your pointer over the graph to display the prices at specific times in the table below the graph.</li>
</ol>
</div>
<div id="pricing-history-command-line" class="section level2">
<h2>Pricing History (command line)</h2>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-spot-price-history.html">Look at the doc</a></p>
<p>In the <code>Terminal</code> use the code below:</p>
<pre class="bash"><code>s=&quot;--start-time 2019-09-15T09:45:00&quot;          # Start-time UTC format
e=&quot;--end-time 2019-10-15T09:45:00&quot;          # End-time UTC format
t=&quot;--instance-types i3.8xlarge&quot;     # Instance-type
aws ec2 describe-spot-price-history $s $e $t &gt; spot.price.txt

# To get the date and time in UTC format in R: format(Sys.time(), &quot;%Y-%m-%dT%H:%M:%S-0400&quot;)</code></pre>
<p>In R, we can easily generate boxplots showing the <strong>spot price statistics by regions and os</strong>:</p>
<pre class="r"><code>require(tidyverse)
readr::read_tsv(
  file = &quot;spot.price.txt&quot;, 
  col_names = c(&quot;REGION&quot;, &quot;INSTANCES&quot;, &quot;OS&quot;, &quot;SPOT_PRICE&quot;, &quot;DATE&quot;), 
  col_types = &quot;_cccnc&quot;
) %&gt;% 
  ggplot2::ggplot(data = ., ggplot2::aes(x = OS, y = SPOT_PRICE)) +
  ggplot2::geom_boxplot() +
  ggplot2::labs(y = &quot;EC2 Spot Price ($)&quot;, x = &quot;OS&quot;, title = &quot;Spot price history&quot;) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 1)) +
  ggplot2::facet_grid(. ~ REGION, scales = &quot;free&quot;)</code></pre>
<p><img src="figures/spot.price.png" width="1181" style="display: block; margin: auto;" /></p>
</div>
<div id="generate-specification-file" class="section level2">
<h2>Generate specification file</h2>
<p>This is a <code>json</code> file to help to get things done faster during the request:</p>
<pre class="bash"><code>touch ec2_radseq_specification.json
nano ec2_radseq_specification.json</code></pre>
<p><strong>Copy/paste in the <code>nano</code> editor:</strong></p>
<pre class="bash"><code>{
  &quot;ImageId&quot;: &quot;ami-09abbab8663956c8a&quot;,
  &quot;InstanceType&quot;: &quot;i3.8xlarge&quot;,
  &quot;KeyName&quot;: &quot;radseq_keys&quot;,
  &quot;NetworkInterfaces&quot;: [
    {
      &quot;DeviceIndex&quot;: 0,
      &quot;SubnetId&quot;: &quot;subnet-0dee7de0007fac906&quot;,
      &quot;Groups&quot;: [&quot;sg-0c6ffc9bb018f9005&quot;],
      &quot;AssociatePublicIpAddress&quot;: true
    }
    ]
}</code></pre>
<p><strong>Write the file and exit:</strong> <code>^x</code> (control-x), <code>y</code> (for yes) and <code>Enter</code> (the keyboard key…)</p>
</div>
<div id="request-the-spot-instance" class="section level2">
<h2>Request the spot instance</h2>
<pre class="bash"><code>z=&quot;--availability-zone-group us-east-1a&quot;      # availability zone
n=&quot;--instance-count 1&quot;                        # number of spot instances
p=&quot;--spot-price 0.99&quot;                          # maximum price 
r=&quot;--type one-time&quot;                           # request type &#39;one-time|persistent&#39;
s=&quot;--launch-specification file://ec2_radseq_specification.json&quot;

aws ec2 request-spot-instances $z $n $p $r $s</code></pre>
<ul>
<li>To get the description of the Spot Instance Request use this command: <code>aws ec2 describe-spot-instance-requests</code></li>
<li>If you need to cancel the request, use the request ID from the description into this command:</li>
</ul>
<pre class="bash"><code>request_id=&quot;your-spot-instance-request-id&quot;
aws ec2 cancel-spot-instance-requests $request_id</code></pre>
<p><strong>Further reading:</strong> <a href="http://ec2-downloads.s3.amazonaws.com/Intro-to-Spot-Instances.pdf">Introduction to spot instances</a></p>
</div>
</div>
<div id="connect" class="section level1">
<h1>Connection with the instance</h1>
<div id="ssh-connection" class="section level2">
<h2>SSH connection</h2>
<p>For the <a href="https://en.wikipedia.org/wiki/SSH_File_Transfer_Protocol">SSH</a> connection to work, provide the information for:</p>
<ul>
<li>the instance public DNS</li>
<li>the path to your key pair generated above</li>
</ul>
<p><strong>Description of your instance</strong></p>
<p>Look in Amazon console for approval of your spot instance request or use the code below to get the description of your instance:</p>
<pre class="bash"><code>aws ec2 describe-instances</code></pre>
<p><strong>Write down the instance public DNS:</strong></p>
<pre class="bash"><code>instance=&quot;ec2-3-85-86-52.compute-1.amazonaws.com&quot;</code></pre>
<p><strong>Write down the instance id</strong></p>
<pre class="bash"><code>instance_id=&quot;i-0f9207c5a0dc76303&quot;</code></pre>
<p>This info is also available in the browser console</p>
<p><a href="https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html">(documentation)</a></p>
<pre class="bash"><code>ssh -i &quot;radseq_keys.pem&quot; ec2-user@$instance</code></pre>
<p>The output:</p>
<pre class="bash"><code>Are you sure you want to continue connecting (yes/no)?
# answer... yes</code></pre>
<ul>
<li>Shows processing units available: <code>nproc</code></li>
<li>For human readable CPU architecture information: <code>lscpu</code></li>
</ul>
</div>
<div id="ebs-volumes" class="section level2">
<h2>EBS volumes</h2>
<p>The <strong>i3.8xlarge</strong> instance comes with 4 SSD storage <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/EBSVolumes.html">(EBS volume)</a>. Use the <code>lsblk</code> command to:</p>
<ul>
<li>view your available disk devices</li>
<li>view their mounting points</li>
<li>to determine the correct device name to use.</li>
</ul>
<p>Note that the output of <code>lsblk</code> removes the <code>/dev/</code> prefix from full device paths.</p>
<div id="format-the-drives" class="section level3">
<h3>Format the drives</h3>
<pre class="bash"><code>sudo mkfs -t ext4 /dev/nvme0n1
sudo mkfs -t ext4 /dev/nvme1n1
sudo mkfs -t ext4 /dev/nvme2n1
sudo mkfs -t ext4 /dev/nvme3n1</code></pre>
</div>
<div id="mount-the-drives" class="section level3">
<h3>Mount the drives</h3>
<div id="ebs-volumes-1" class="section level4">
<h4>EBS volumes</h4>
<pre class="bash"><code>sudo mount /dev/nvme0n1 /media/ebs_1
sudo mount /dev/nvme1n1 /media/ebs_2
sudo mount /dev/nvme2n1 /media/ebs_3
sudo mount /dev/nvme3n1 /media/ebs_4</code></pre>
</div>
<div id="s3-bucket" class="section level4">
<h4>s3 bucket</h4>
<p>You need to <strong>edit s3fs</strong>, the software used to mount your s3 bucket to your Amazon instance.</p>
<pre class="bash"><code>sudo nano /etc/passwd-s3fs                # edit the personal file used by s3fs
s3-bucket-name:AccessKey:SecretKey        # add the information keeping &quot; : &quot;
crtl-o                                    # to write the change to the file
crtl-x                                    # to exit nano editor
chmod 640 /etc/passwd-s3fs                # set permissions (might need sudo)
chown ec2-user:ec2-user /etc/passwd-s3fs  # change ownership
s3fs -o allow_other gbs_data /media/s3/   # mount s3 bucket &#39;gbs_data&#39;</code></pre>
<p><strong>test your mounting installation:</strong></p>
<pre class="bash"><code>sudo touch /media/s3/testing # will save the file &#39;testing&#39; to your bucket
ls -l /media/s3 # will show the content of your bucket and your testing file</code></pre>
<p>See <a href="https://github.com/s3fs-fuse/s3fs-fuse/wiki/Fuse-Over-Amazon">s3fs documentation</a></p>
<p><strong>s3 drive mount on instance reboot</strong></p>
<ol style="list-style-type: decimal">
<li>Creates a new batch file called <strong>automount-s3</strong> containing the s3fs command to mount your S3 bucket:</li>
<li>Move the newly created file to location <code>/usr/sbin</code></li>
<li>Change ownership of file to <code>root:root</code></li>
<li>Change permissions of file to <code>+x</code> (executable)</li>
<li>Reboot your system and check that the script actually works</li>
</ol>
<pre class="bash"><code>echo &quot;s3fs bucket_name -o allow_other /media/s3&quot; &gt;&gt; automount-s3
sudo mv automount-s3 /usr/sbin
sudo chown root:root /usr/sbin/automount-s3
sudo chmod +x /usr/sbin/automount-s3
sudo reboot   # to reboot the EC2 Instance</code></pre>
<p>After reboot is complete, connect to your instance and run the executable file</p>
<pre class="bash"><code>instance=&quot;ec2-54-87-41-96.compute-1.amazonaws.com&quot;        # public DNS
keypair_path=&quot;radseq_keypair.pem&quot;                         # path to your key pair
ssh -i $keypair_path ec2-user@$instance                   # to start the connection
sudo -i  # become root and keep sudo privilege
sudo /usr/sbin/automount-s3
df -h # see if disk is mounted and command successful</code></pre>
</div>
</div>
<div id="additionnal-disk-space" class="section level3">
<h3>Additionnal disk space:</h3>
<p>If you need more disk space:</p>
<ul>
<li><a href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-creating-volume.html">EBS drive</a></li>
<li><a href="http://aws.amazon.com/ebs/pricing/">pricing</a></li>
</ul>
<p><strong>Create a 1 TB EBS drive:</strong></p>
<pre class="bash"><code>SIZE=1000
TYPE=standard
ZONE=us-east-1a

aws ec2 create-volume --size $SIZE --volume-type $TYPE --availability-zone $ZONE</code></pre>
<p><strong>Describe the volumes you have:</strong></p>
<pre class="bash"><code>aws ec2 describe-volumes</code></pre>
<p><strong>Attach</strong> the 1TB EBS drive to a running or stopped instance:</p>
<pre class="bash"><code>I=&quot;--instance-id i-611a8642&quot;            # ID of the instance
EBS=&quot;--volume-id vol-e859faa4&quot;          # ID of Amazon EBS volume
D=&quot;--device /dev/xvdf&quot;                  # the device name

aws ec2 attach-volume $I $EBS $D</code></pre>
<p>After the 1 TB EBS volume is attached, you’ll have to <strong>format</strong>, <strong>make a mounting point</strong> and <strong>mount the EBS volume</strong>.</p>
<pre class="bash"><code>sudo mkfs -t ext4 /dev/xvdd          # format
sudo mkdir /media/ebs_3              # create mounting point
sudo mount /dev/xvdd /media/ebs_3    # mount</code></pre>
<p><strong>Detach</strong> a volume from an instance use these commands:</p>
<pre class="bash"><code>I=&quot;--instance-id i-611a8642&quot;            # ID of the instance
EBS=&quot;--volume-id vol-e859faa4&quot;          # ID of Amazon EBS volume
D=&quot;--device /dev/xvdf&quot;                  # the device name

aws ec2 detach-volume $I $EBS $D</code></pre>
<p><strong>Note: the difference between S3 and EBS drives:</strong></p>
<ol style="list-style-type: decimal">
<li>EBS can only be used with EC2 instances while S3 can be used outside EC2</li>
<li>EBS appears as a mountable volume while the S3 requires software to read and write data</li>
<li>EBS can accommodate a smaller amount of data than S3</li>
<li>EBS can only be used by one EC2 instance at a time while S3 can be used by multiple instances</li>
<li>S3 typically experiences write delays while EBS does not</li>
</ol>
</div>
</div>
<div id="terminate-your-instance" class="section level2">
<h2>Terminate your instance</h2>
<p>Don’t forget to terminate your instance when your analysis are completed!</p>
<pre class="bash"><code>aws ec2 terminate-instances --instance-ids $instance_id # modify to match your instance id</code></pre>
</div>
<div id="tags" class="section level2">
<h2>Tags</h2>
<p>Add a <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/Using_Tags.html">tag</a> for bookkeeping</p>
<pre class="bash"><code>resource=&quot;--resources ami-55fa8e3c&quot; # the AWS-assigned ID to tag
tag=&quot;--tags radse=1&quot; # key=value of the tag

aws ec2 create-tags $resource $tag</code></pre>
</div>
</div>
<div id="stacks" class="section level1">
<h1>Run Stacks</h1>
<p>Now, I guess you can’t wait to try <a href="http://catchenlab.life.illinois.edu/stacks/">Stacks</a>… Check that everything is working properly:</p>
<pre class="bash"><code>populations # to test your installation!</code></pre>
<p>You might have to give the full path if the above command doesn’t work:</p>
<pre class="bash"><code>/usr/local/bin/populations</code></pre>
<p><strong>You are ready to start analyzing your RADseq data!</strong></p>
<ol style="list-style-type: decimal">
<li>Output your analysis directly in one of the EBS volume</li>
<li>When your analysis are completed, compress your folders using: <code>tar -czvf stacks_output.tar.gz path/of/stacks_output/files</code> and transfer the results to your <strong>s3 bucket</strong> to have access to your data when the instance is shut down.</li>
</ol>
</div>
<div id="userstudio" class="section level1">
<h1>Use RStudio</h1>
<p>You can use our public <a href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html">AMI</a> already loaded with RADseq software + <a href="https://cran.r-project.org">R</a> + <a href="https://www.rstudio.com">RStudio</a>.</p>
<p>To access RStudio server you need to configure the security group that will allow commutication with <strong>port 8787</strong>. This is explain in the <strong>security group section above</strong>.</p>
<p>In your favourite web browser you can type the public ip adress of your instance with the port <code>8787</code> at the end:</p>
<pre class="bash"><code>54.172.183.211:8787
# These credentials will be asked:
# Username: rstudio
# Password: rstudio</code></pre>
</div>
<div id="ami" class="section level1">
<h1>Creating your own AMI</h1>
<p><strong>Time required: ~ 15min</strong>.</p>
<p>Spot instances can’t be used to generate AMI because the instance needs to be stopped easily and re-started. This is best accomplished with a <a href="https://aws.amazon.com/ec2/purchasing-options/">On-Demand Instances</a>.</p>
<p>For the example, let’s use <a href="https://aws.amazon.com/amazon-linux-2/">Amazon Linux 2 AMI (HVM)</a>. The image comes with 8GB of disk space and will likely not be enough to install most genomic software you want. To increase the size we need the information called <em>block-device-mappings</em>.</p>
<p>ID of the image:</p>
<pre class="bash"><code>ami=&quot;ami-0b69ea66ff7391e80&quot;</code></pre>
<p>Additionnal info required (generated above):</p>
<pre class="bash"><code>subnet=&quot;subnet-0dee7de0007fac906&quot;
sg=&quot;sg-0c6ffc9bb018f9005&quot;</code></pre>
<div id="increase-disk-space" class="section level2">
<h2>Increase disk space</h2>
<p>Get the info on <em>block-device-mappings</em>:</p>
<pre class="bash"><code>aws ec2 describe-images --image-id $ami --output json</code></pre>
<p>We need:</p>
<ul>
<li>DeviceName: <code>/dev/xvda</code></li>
<li>SnapshotId: <code>snap-0e4c15b8cba3e8ae6</code></li>
<li>VolumeType: <code>gp2</code></li>
</ul>
<p>Generate a <code>json</code> file to help to get things done faster during the request:</p>
<pre class="bash"><code>touch ec2_radseq_mapping.json
nano ec2_radseq_mapping.json</code></pre>
<p>Modify values below based on the image description obtained above, copy/paste in the <code>nano</code> editor:</p>
<pre class="bash"><code>[
 {
    &quot;DeviceName&quot;: &quot;/dev/xvda&quot;,
    &quot;Ebs&quot;: {
        &quot;DeleteOnTermination&quot;: true,
        &quot;SnapshotId&quot;: &quot;snap-0e4c15b8cba3e8ae6&quot;,
        &quot;VolumeSize&quot;: 30,
        &quot;VolumeType&quot;: &quot;gp2&quot;,
        &quot;Encrypted&quot;: false
    }
}
]</code></pre>
<p>To exit <code>nano</code>: with your keyboard: <code>^x</code> (control-x), <code>y</code> and <code>enter</code></p>
</div>
<div id="request-an-instance" class="section level2">
<h2>Request an instance</h2>
<pre class="bash"><code>t=&quot;--instance-type i3.8xlarge&quot;                           # instance type
k=&quot;--key-name radseq_keys&quot;                               # name of YOUR keypair
n=&quot;--count 1&quot;                                            # number of instances
i=&quot;--associate-public-ip-address&quot;                        # add a public IP address
aws ec2 run-instances --image-id $ami $t $z $k $n $i --security-group-ids $sg --subnet-id $subnet --block-device-mappings file://ec2_radseq_mapping.json</code></pre>
</div>
<div id="start-ssh-connection" class="section level2">
<h2>Start <a href="http://en.wikipedia.org/wiki/Secure_Shell">SSH</a> connection</h2>
<pre class="bash"><code>aws ec2 describe-instances</code></pre>
<p>Write down the instance public DNS (starts with <code>ec2-</code> and ends with <code>compute-1.amazonaws.com</code>) and id (starts with <code>i-</code>:</p>
<pre class="bash"><code>instance=&quot;ec2-54-159-198-106.compute-1.amazonaws.com&quot;
instance_id=&quot;i-0f9207c5a0dc76303&quot;</code></pre>
<pre class="bash"><code>ssh -i &quot;radseq_keys.pem&quot; ec2-user@$instance</code></pre>
<p>To this question (if you have it):</p>
<pre class="bash"><code>Are you sure you want to continue connecting (yes/no)?</code></pre>
<p>Answer: <code>yes</code></p>
<p>Become root user and keep your sudo privilege active throughout the session:</p>
<pre class="bash"><code>sudo -i</code></pre>
</div>
<div id="updates.." class="section level2">
<h2>Updates..</h2>
<p>For librairies, software/dependencies and system updates, run the following commands:</p>
<pre class="bash"><code>yum update -y</code></pre>
<p>Note that <strong>Fedora/Linux</strong> distros uses <code>yum</code> while <strong>Debian/Ubuntu</strong> distros uses <code>apt-get</code> natively.</p>
<p>Allow access to the Extra Packages for Enterprise Linux (<a href="https://fedoraproject.org/wiki/EPEL">EPEL</a>):</p>
<pre class="bash"><code>sudo yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm</code></pre>
<p>Also install these dependencies:</p>
<pre class="bash"><code>yum install -y automake autoconf bzip2-devel curl-devel xerces-c expat-devel fuse fuse-devel freetype* gcc gcc-c++ gcc-gfortran gettext gettext-devel git gsl-devel java-1.8.0-openjdk-devel libcurl-devel libpng-devel libstdc++-devel libX11-devel libxml2-devel libXt-devel libtool mailcap mesa-libGLU-devel mysql mysql-devel numpy openssh-* openssl-devel python-devel python-magic python-pip readline-devel sqlite-devel subversion swig texinfo-tex trickle unixODBC-devel unzip zlib-devel</code></pre>
</div>
<div id="ebs-drives" class="section level2">
<h2>EBS drives</h2>
<p>The <strong>i3.8xlarge</strong> instance comes with 4 SSD storage <a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/EBSVolumes.html">(EBS volume)</a>.</p>
<p>Before accessing those drives you need to:</p>
<ul>
<li>Create mounting directories</li>
<li>Format the drives</li>
<li>Mount the drives on the instance</li>
</ul>
<div id="create-the-directories" class="section level3">
<h3>Create the directories</h3>
<pre class="bash"><code>sudo mkdir /media/ebs_1
sudo mkdir /media/ebs_2
sudo mkdir /media/ebs_3
sudo mkdir /media/ebs_4
# give proper permission:
sudo chown -R ec2-user:root /media/ebs_1
sudo chown -R ec2-user:root /media/ebs_2
sudo chown -R ec2-user:root /media/ebs_3
sudo chown -R ec2-user:root /media/ebs_4</code></pre>
</div>
<div id="format" class="section level3">
<h3>Format</h3>
<p>Use the <code>lsblk</code> command to:</p>
<ul>
<li>view your available disk devices</li>
<li>view their mounting points</li>
<li>to determine the correct device name to use.</li>
</ul>
<p>Note that the output of <code>lsblk</code> removes the <code>/dev/</code> prefix from full device paths.</p>
<pre class="bash"><code>sudo mkfs -t ext4 /dev/nvme0n1
sudo mkfs -t ext4 /dev/nvme1n1
sudo mkfs -t ext4 /dev/nvme2n1
sudo mkfs -t ext4 /dev/nvme3n1</code></pre>
</div>
<div id="mount" class="section level3">
<h3>Mount</h3>
<pre class="bash"><code>sudo mount /dev/nvme0n1 /media/ebs_1
sudo mount /dev/nvme1n1 /media/ebs_2
sudo mount /dev/nvme2n1 /media/ebs_3
sudo mount /dev/nvme3n1 /media/ebs_4</code></pre>
</div>
</div>
<div id="s3-bucket-1" class="section level2">
<h2>s3 bucket</h2>
<p>If you plan on using a s3 bucket, you need to install s3fs tool and configure it:</p>
<div id="install-s3fs" class="section level3">
<h3>Install s3fs</h3>
<pre class="bash"><code>cd /home/ec2-user
git clone https://github.com/s3fs-fuse/s3fs-fuse.git
cd s3fs-fuse
./autogen.sh
./configure --prefix=/usr --with-openssl &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R s3fs-fuse*</code></pre>
</div>
<div id="configure-s3fs" class="section level3">
<h3>Configure s3fs</h3>
<p>Open fuse configuration file:</p>
<pre class="bash"><code>sudo nano /etc/fuse.conf</code></pre>
<ul>
<li>Modify the line that starts with <code>#user_allow_other</code>:</li>
<li>Uncomment the line by removing the <code>#</code> in front of <code>user_allow_other</code></li>
<li>Exit: <code>^x</code>, Answer <code>Y</code> to: <code>Save modified buffer? (Answering &quot;No&quot; will DISCARD changes</code> and <code>enter</code></li>
<li>Change the file permission: <code>sudo chmod 640 /etc/fuse.conf</code></li>
</ul>
<p>Allow s3fs to use your s3 bucket:</p>
<ul>
<li>you need your Amazon <strong>Access Key</strong> and the <strong>Secret Key</strong>.</li>
<li>Generate a password file to access the s3 bucket:</li>
</ul>
<pre class="bash"><code>sudo touch /etc/passwd-s3fs
sudo nano /etc/passwd-s3fs</code></pre>
<ul>
<li>Enter in the file <code>s3 bucket name:aws_access_key_id:aws_secret_access_key</code> e.g. if bucket name is radseq, this is what’s in the file:</li>
</ul>
<pre class="bash"><code>radseq:TKUDJHLS3YPQSHWIKDBF:sLLDoi9uSBoBD0g8ECH3ZDTL9Onio2Ky5CdLm87C</code></pre>
<ul>
<li>Save and Exit: <code>^x</code>, answer <code>y</code>, hit <code>enter</code></li>
<li>Change file permission:</li>
</ul>
<pre class="bash"><code>sudo chmod 640 /etc/passwd-s3fs
sudo chown ec2-user:root /etc/passwd-s3fs</code></pre>
</div>
<div id="create-the-directory" class="section level3">
<h3>Create the directory</h3>
<pre class="bash"><code>sudo mkdir -p /media/s3
sudo chown -R ec2-user:root /media/s3</code></pre>
</div>
<div id="mount-s3-bucket" class="section level3">
<h3>Mount s3 bucket</h3>
<pre class="bash"><code>s3fs -o allow_other radseq /media/s3</code></pre>
</div>
<div id="useful" class="section level3">
<h3>Useful</h3>
<ul>
<li>To <strong>unmount</strong> a drive use <code>sudo umount /media/s3/</code></li>
<li>To see if your EBS drives are mounted use this command: <code>df -h</code></li>
</ul>
<p><strong>Congratulations! You now have a mounted s3 bucket and 4 EBS volumes mounted on your instance.</strong></p>
</div>
</div>
<div id="next-steps" class="section level2">
<h2>Next steps</h2>
<p>At this point your instance is ready to work. If your ok with what was installed and just want to go with that, the next step is <a href="#save_ami">saving your AMI</a>. You can also continue installing software before saving your ami. Below are 2 additional and optinal sections:</p>
<ul>
<li>Install <a href="#RADseq_software">RADseq sofware</a></li>
<li>Install <a href="#rstudio">R and RStudio</a></li>
</ul>
</div>
<div id="save_ami" class="section level2">
<h2>Save your AMI</h2>
<p><strong>Time required: &lt; 5min</strong></p>
<div id="remove-sensitive-data" class="section level3">
<h3>Remove sensitive data</h3>
<p>If the goal is to share your AMI (make it available to the public) always remove sensitive data:</p>
<ul>
<li><strong>Delete the shell history</strong>: the shell history may contain your secret access key or other private info that are not intended to be shared with users of your AMI.</li>
</ul>
<p>As an example, the following command can help locate the root user and other users’ shell history files on disk and delete them, when run as root:</p>
<pre class="bash"><code>find /root/.*history /home/*/.*history -exec rm -f {} \;</code></pre>
<ul>
<li><strong>Remove your <code>s3cfg configuration file</code> and <code>s3fs password</code>:</strong></li>
</ul>
<pre class="bash"><code>sudo rm /root/.s3cfg /etc/passwd-s3fs</code></pre>
</div>
<div id="unmount-drives" class="section level3">
<h3>Unmount drives</h3>
<pre class="bash"><code># From your instance:
sudo umount /media/s3/
sudo umount /media/ebs_1/
sudo umount /media/ebs_2/
sudo umount /media/ebs_3/
sudo umount /media/ebs_4/</code></pre>
</div>
<div id="stop-your-instance" class="section level3">
<h3>Stop your instance</h3>
<pre class="bash"><code>#On your computer: 
aws ec2 stop-instances --instance-ids $instance_id</code></pre>
</div>
<div id="generate-a-private-ami" class="section level3">
<h3>Generate a private AMI</h3>
<pre class="bash"><code># On your computer
aws ec2 create-image --instance-id $instance_id --name &quot;radseq&quot; --description &quot;radseq analysis image&quot;</code></pre>
<p><a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/creating-an-ami-ebs.html">Documentation to create an Amazon EBS-backed AMI</a></p>
<p>It can take several minutes to generate the image. To get the description of your image, use this command:</p>
<pre class="bash"><code>aws ec2 describe-images --executable-users all</code></pre>
<p>Write down the ami id:</p>
<pre class="r"><code>ami=&quot;ami-09abbab8663956c8a&quot;</code></pre>
<p>Congratulation you now have a <strong>private AMI</strong> associated with your Amazon AWS account!</p>
<p><a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/creating-an-ami-ebs.html">Documentation</a></p>
</div>
<div id="make-your-ami-public" class="section level3">
<h3>Make your AMI public</h3>
<p>To make your AMI public, modify the AMI’s launch permissions by adding <code>all</code> to the launch permission attribute:</p>
<pre class="bash"><code>aws ec2 modify-image-attribute --image-id $ami --launch-permission &quot;Add=[{Group=all}]&quot;</code></pre>
<p>To verify the lauch permissions of the AMI:</p>
<pre class="bash"><code>aws ec2 describe-image-attribute --image-id $ami --attribute launchPermission</code></pre>
<p><a href="https://docs.aws.amazon.com/en_pv/AWSEC2/latest/UserGuide/sharingamis-intro.html">Documentation for sharing AMI</a>.</p>
</div>
<div id="useful-1" class="section level3">
<h3>Useful</h3>
<div id="create-a-snapshots" class="section level4">
<h4>Create a <a href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-creating-snapshot.html">snapshots</a></h4>
<p>From a <strong>stopped instance’s</strong> root volume:</p>
<pre class="bash"><code>aws ec2 create-snapshot --volume-id vol-1234567890abcdef0 --description &quot;for RADseq analysis image&quot;</code></pre>
</div>
<div id="re-start-your-instance" class="section level4">
<h4>Re-start your instance</h4>
<pre class="bash"><code>aws ec2 start-instances --instance-ids $instance_id</code></pre>
</div>
</div>
</div>
</div>
<div id="RADseq_software" class="section level1">
<h1>Install GBS/RADseq software</h1>
<p>Below are just a few software you can install and run on Amazon EC2. <strong>Time required: ~ 10min</strong>.</p>
<div id="path" class="section level2">
<h2>PATH</h2>
<p><strong>1. Make sure you have <code>/usr/local/bin</code> in your <code>PATH</code>:</strong></p>
<pre class="bash"><code>sudo nano /home/ec2-user/.bash_profile</code></pre>
<p>Add <code>/usr/local/bin</code> at the end of the line starting with <code>PATH</code>.</p>
<p><strong>2. Add <code>LD_LIBRARY_PATH</code></strong></p>
<p>Also add this line in the file:</p>
<pre class="bash"><code>export LD_LIBRARY_PATH=/usr/local/lib/</code></pre>
<p><strong>Save and Exit</strong>: <code>^x</code>, answer <code>y</code>, hit <code>enter</code></p>
<pre class="bash"><code>source /home/ec2-user/.bash_profile</code></pre>
<p>Inside my <code>.bash_profile</code> file, it look like:</p>
<pre class="bash"><code># .bash_profile

# Get the aliases and functions
if [ -f ~/.bashrc ]; then
    . ~/.bashrc
fi

# User specific environment and startup programs

PATH=$PATH:$HOME/.local/bin:$HOME/bin:/usr/local/bin
export PATH
export LD_LIBRARY_PATH=/usr/local/lib/</code></pre>
</div>
<div id="bayescan" class="section level2">
<h2><a href="https://www.cmpg.iee.unibe.ch/services/software/bayescan/index_eng.html">BayeScan</a></h2>
<ul>
<li>Download</li>
<li>Move and rename the binary to <code>/usr/local/bin</code></li>
<li>Set permission</li>
<li>Remove downloaded file/folder</li>
</ul>
<pre class="bash"><code>wget http://cmpg.unibe.ch/software/BayeScan/files/BayeScan2.1.zip
unzip BayeScan2.1.zip
sudo cp BayeScan2.1/binaries/BayeScan2.1_linux64bits /usr/local/bin/bayescan
sudo chmod 777 /usr/local/bin/bayescan 
sudo rm -R BayeScan*
bayescan # to test</code></pre>
</div>
<div id="vcftools" class="section level2">
<h2><a href="https://vcftools.github.io/index.html">VCFtools</a></h2>
<pre class="bash"><code>git clone https://github.com/vcftools/vcftools.git
cd vcftools
./autogen.sh
./configure &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R vcftools
vcftools # to test</code></pre>
</div>
<div id="plink" class="section level2">
<h2><a href="https://www.cog-genomics.org/plink/2.0/">PLINK</a></h2>
<pre class="bash"><code>wget http://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_20191020.zip
unzip plink2* -d plink
sudo cp plink/plink2 /usr/local/bin/plink # password required
sudo rm -R plink* # to remove the folder and zip file
plink # to test</code></pre>
</div>
<div id="fastsimcoal2" class="section level2">
<h2><a href="https://www.cmpg.iee.unibe.ch/services/software/fastsimcoal2/index_eng.html">fastsimcoal2</a></h2>
<pre class="bash"><code>wget http://cmpg.unibe.ch/software/fastsimcoal2/downloads/fsc26_linux64.zip
unzip fsc26_linux64.zip
sudo mv fsc26_linux64/fsc26 /usr/local/bin/fsc26
sudo chmod 777 /usr/local/bin/fsc26
sudo rm -R fsc26*
fsc26 # to test</code></pre>
</div>
<div id="stacks-1" class="section level2">
<h2><a href="http://catchenlab.life.illinois.edu/stacks">stacks</a></h2>
<p><strong>To install stacks, you’ll need to install:</strong></p>
<ul>
<li><a href="https://github.com/lh3/bwa">BWA</a>: Burrow-Wheeler Aligner</li>
<li><a href="https://github.com/samtools/htslib">htslib</a>: C library for high-throughput sequencing data formats.</li>
<li><a href="http://www.htslib.org/download/">SAMtools</a></li>
<li><a href="https://github.com/samtools/bcftools">BCFtools</a></li>
</ul>
<div id="burrow-wheeler-aligner-bwa" class="section level3">
<h3><a href="https://github.com/lh3/bwa">Burrow-Wheeler Aligner, BWA</a></h3>
<pre class="bash"><code>git clone https://github.com/lh3/bwa.git
cd bwa
make -j99
sudo mv bwa /usr/local/bin/bwa
cd ..
sudo rm -R bwa</code></pre>
</div>
<div id="htslib" class="section level3">
<h3><a href="https://github.com/samtools/htslib">htslib</a></h3>
<pre class="bash"><code>git clone https://github.com/samtools/htslib.git
cd htslib
autoheader
autoconf
./configure --prefix=/usr/local &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R htslib*</code></pre>
<p>Note that this will install <strong>htslib</strong> files in: <code>/usr/local/lib</code>, <code>/usr/local/include</code>, <code>/usr/local/bin</code>, <code>/usr/local/share</code></p>
<p>To change where executables are installed use: <code>./configure --prefix=/where/to/install</code> (remember to add this directory to your <code>$PATH:</code>).</p>
</div>
<div id="samtools" class="section level3">
<h3><a href="http://www.htslib.org/download/">SAMtools</a></h3>
<pre class="bash"><code>git clone https://github.com/samtools/samtools.git
cd samtools*
autoheader
autoconf -Wno-syntax 
./configure --with-htslib=/usr/local &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R samtools*
samtools</code></pre>
</div>
<div id="bcftools" class="section level3">
<h3><a href="https://github.com/samtools/bcftools">BCFtools</a></h3>
<pre class="bash"><code>git clone https://github.com/samtools/bcftools.git
cd bcftools*
autoheader
autoconf
./configure &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R bcftools*</code></pre>
</div>
<div id="download-and-install-stacks" class="section level3">
<h3>Download and install <a href="http://catchenlab.life.illinois.edu/stacks">stacks</a></h3>
<pre class="bash"><code>wget http://catchenlab.life.illinois.edu/stacks/source/stacks-2.41.tar.gz # you want to use version 1.48 instead? Just change to the version number you want.
tar -xvf stacks*
cd stacks-*
./configure &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R stacks* # to remove the folder and gz file
populations # to test your installation!</code></pre>
</div>
</div>
<div id="save-your-ami" class="section level2">
<h2><a href="#save_ami">Save your AMI</a></h2>
</div>
</div>
<div id="rstudio" class="section level1">
<h1>Install R and RStudio Server</h1>
<p><strong>Time required: ~ 45min</strong>.</p>
<p>Amazon always as outdated R version. Consequently, you need to install the latest version:</p>
<div id="install-r-from-source" class="section level2">
<h2>Install R from source</h2>
<pre class="bash"><code>cd /home/ec2-user
wget https://cran.r-project.org/src/base/R-latest.tar.gz
tar -xvf R-latest.tar.gz
cd R-*
./configure --with-x=yes --enable-R-shlib=yes --with-cairo=yes &amp;&amp; make -j99 &amp;&amp; sudo make install #take a coffee break!
cd ..
sudo rm -R R-*</code></pre>
<ul>
<li>to test the installation, type <code>R</code> in the Terminal</li>
<li>to quit R: <code>q()</code></li>
<li><code>Save workspace image? [y/n/c]:</code> answer <code>n</code></li>
</ul>
</div>
<div id="update-your-path" class="section level2">
<h2>Update your PATH</h2>
<p>in the file: <code>/home/ec2-user/.bash_profile</code>, using <code>sudo nano /home/ec2-user/.bash_profile</code> as shown above or a text editor, add these lines:</p>
<pre class="bash"><code>export PATH=/usr/local/bin/R:$PATH
export RSTUDIO_WHICH_R=/usr/local/bin/R</code></pre>
<p>Don’t forget to refresh with:</p>
<pre class="bash"><code>source /home/ec2-user/.bash_profile</code></pre>
</div>
<div id="install-rstudio-server" class="section level2">
<h2>Install RStudio Server</h2>
<p>This will allow you to use RStudio throygh your favorite internet browser…</p>
<pre class="bash"><code>cd /home/ec2-user
wget https://download2.rstudio.org/server/centos6/x86_64/rstudio-server-rhel-1.2.5001-x86_64.rpm
sudo yum install -y rstudio-server-rhel-1.2.5001-x86_64.rpm
sudo rm -R rstudio-server*</code></pre>
<p>Configure:</p>
<pre class="bash"><code>sudo adduser rstudio</code></pre>
<p>Change password:</p>
<p>If you’re not already at the root level type: <code>sudo -i</code></p>
<pre class="bash"><code>passwd rstudio</code></pre>
<ul>
<li>e.g. you could use <strong>radseq2019</strong> or any other password you want.</li>
<li>you’ll have to enter it twice</li>
</ul>
<p>Grant user access to the rstudio home folder:</p>
<pre class="bash"><code>sudo chmod -R ugo+rwx /home/rstudio/</code></pre>
<p>We need to set the library path for RStudio:</p>
<pre class="bash"><code>sudo nano /etc/rstudio/rserver.conf</code></pre>
<p>Add this line:</p>
<pre class="bash"><code>rsession-ld-library-path=/usr/lib64/:/usr/local/lib/</code></pre>
<p><strong>Save and Exit</strong>: <code>^x</code>, answer <code>y</code>, hit <code>enter</code></p>
<p>To test that <strong>RStudio server</strong> works, go to your browser and use the public IP of the instance followed by <code>:8787</code></p>
<p>Get the public ip:</p>
<pre class="bash"><code># from the Terminal ON YOUR COMPUTER or a new Terminal window:
aws ec2 describe-instances</code></pre>
<p>Or from Amazon EC2 console…</p>
<p>In the browser:</p>
<pre class="bash"><code>54.159.198.106:8787</code></pre>
<p>You should see: <img src="figures/rstudio_login.png" width="320" style="display: block; margin: auto;" /></p>
<p><strong>Enter the credentials:</strong></p>
<ul>
<li><strong>Username:</strong> rstudio</li>
<li><strong>Password:</strong>radseq2019</li>
</ul>
<p>When everything is configured properly, RStudio running in your browser should look like this: <img src="figures/rstudio_server.png" width="708" style="display: block; margin: auto;" /></p>
<p><strong>Restarting RStudio Server</strong> If you ever need to restart the server:</p>
<pre class="bash"><code>sudo rstudio-server restart</code></pre>
</div>
<div id="r-environment" class="section level2">
<h2>R environment</h2>
<p><strong>1. Create a <code>.Renviron</code> file</strong>:</p>
<pre class="bash"><code>#In Terminal
cd /home/ec2-user
touch .Renviron
sudo nano /home/ec2-user/.Renviron</code></pre>
<p><strong>2. Inside <code>/home/ec2-user/.Renviron</code>, add these lines</strong>:</p>
<pre class="bash"><code>PKG_CONFIG_PATH=&quot;/usr/local/lib/pkgconfig&quot;
LD_LIBRARY_PATH=&quot;/usr/local/lib/:/usr/lib64:/usr/local/lib64/R/lib::/lib:/usr/local/lib64:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.222.b10-0.amzn2.0.1.x86_64/jre/lib/amd64/server&quot;
INCLUDE_PATH=&quot;/usr/include/sys/stat.h:/usr/include/sys/signal.h&quot;</code></pre>
<p><strong>Save and Exit <code>nano</code> with</strong>: <code>^x</code>, answer <code>y</code>, hit <code>enter</code></p>
<p>You could also do all this inside RStudio on your browser:</p>
<ul>
<li>going into the <em>File</em> menu</li>
<li>clicking to <em>New File</em> and then by selecting <em>Text File</em>…</li>
<li>but it’s much faster with the command line…</li>
</ul>
<p><strong>3. make available for users: <code>ec2-user</code> and <code>rstudio</code></strong></p>
<p>Because you might want to use R directly from the Terminal or inside RStudio…</p>
<pre class="bash"><code>sudo cp /home/ec2-user/.Renviron /home/rstudio/.Renviron</code></pre>
<p><strong>4. Permissions</strong>:</p>
<pre class="bash"><code>chown ec2-user:rstudio /home/ec2-user/.Renviron
chown ec2-user:rstudio /home/rstudio/.Renviron
chmod 777 /home/ec2-user/.Renviron
chmod 777 /home/rstudio/.Renviron</code></pre>
</div>
<div id="install-packages" class="section level2">
<h2>Install Packages</h2>
<p>The next steps are totally up to you.</p>
<div id="dependencies" class="section level3">
<h3>Dependencies</h3>
<p>Note that for Linux using <code>radiator</code>, <code>dartR</code>, <code>adegenet</code>or any packages that rely on the <code>sf</code> package… it’s more complicated and you first need to install the dependencies through the <strong>terminal</strong>:</p>
<p><strong>udunits</strong></p>
<pre class="bash"><code>cd /home/ec2-user/
curl -O ftp://ftp.unidata.ucar.edu/pub/udunits/udunits-2.2.26.tar.gz
tar -xvf udunits*
cd udunits*
./configure &amp;&amp; make -j99 &amp;&amp; sudo make install
sudo ldconfig
cd ..
sudo rm -R udunits*</code></pre>
<p>Using <code>sudo nano /home/ec2-user/.bash_profile</code>, add these lines:</p>
<pre class="bash"><code>export UDUNITS2_XML_PATH=&quot;/usr/local/share/udunits/udunits2.xml&quot;</code></pre>
<p>Refresh:</p>
<pre class="bash"><code>source /home/ec2-user/.bash_profile</code></pre>
<p><a href="https://proj.org/about.html">proj</a></p>
<pre class="bash"><code>wget http://download.osgeo.org/proj/proj-6.1.0.tar.gz
tar zxf proj*
cd proj*
./configure --libdir=/usr/lib64 &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R proj*</code></pre>
<p><a href="https://gdal.org">GDAL</a></p>
<pre class="bash"><code>wget http://download.osgeo.org/gdal/CURRENT/gdal-3.0.1.tar.gz
tar zxf gdal*
cd gdal*
./configure --prefix=/usr/local --libdir=/usr/lib64 --with-proj=/usr/local &amp;&amp; make -j99 &amp;&amp; sudo make install #time for coffee...
cd ..
sudo rm -R gdal*</code></pre>
<p><a href="https://trac.osgeo.org/geos">GEOS</a></p>
<pre class="bash"><code>git clone https://git.osgeo.org/gitea/geos/geos.git
cd geos*
./autogen.sh
./configure --libdir=/usr/lib64 &amp;&amp; make -j99 &amp;&amp; sudo make install
cd ..
sudo rm -R geos*</code></pre>
<pre class="bash"><code>source /home/ec2-user/.bash_profile</code></pre>
<p><strong>After those dependencies, I usually go over my browser in RStudio and do these:</strong></p>
<pre class="r"><code>if (!require(&quot;devtools&quot;)) install.packages(&quot;devtools&quot;) # quite long on Linux
install.packages(&quot;tidyverse&quot;) # long
install.packages(&quot;adegenet&quot;) # not too bad
devtools::install_github(&quot;thierrygosselin/grur&quot;) # quite long on Linux
devtools::install_github(&quot;thierrygosselin/assigner&quot;)
# No need to install radiator because it&#39;s a dependency or grur and assigner...à
install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;SeqVarTools&quot;)
BiocManager::install(&quot;SNPRelate&quot;)</code></pre>
<p>If you want to install <code>dartR</code> follow these instructions, because the direct CRAN install won’t work:</p>
<pre class="r"><code>BiocManager::install(&quot;qvalue&quot;)
devtools::install_github(&quot;green-striped-gecko/PopGenReport&quot;)
devtools::install_github(&quot;green-striped-gecko/dartR&quot;)</code></pre>
<p><strong>My vignette <a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html">RADseq genomics in R</a>, as sections on</strong>:</p>
<ul>
<li><a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html#Installation_problems">installation problems</a>.</li>
<li><a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html#specific-librairies-installation">specific libraries installation</a>.</li>
</ul>
</div>
</div>
<div id="save-your-ami-1" class="section level2">
<h2><a href="#save_ami">Save your AMI</a></h2>
</div>
</div>
<div id="usefullink" class="section level1">
<h1>Useful</h1>
<div id="time" class="section level2">
<h2>Time</h2>
<p>To know how much time was taken to execute a command, or shell script, or any other program. Use <code>time</code> in front of the command.</p>
</div>
<div id="screen" class="section level2">
<h2>Screen</h2>
<p>If you ever accidentally close the Terminal application or get disconnected over SSH, the state of all your processes will be lost unless you use the screen commands!</p>
<p><strong>Important Screen commands:</strong></p>
<pre class="bash"><code>screen -h               # Display options
screen -list            # List open screens
screen -r sessionID     # Reatach to screen session with sessionID
screen -S sessionID     # Create new screen session with name &#39;sessionID&#39;
screen -D sessionID     # Force detach session (if cannot re-attach)
ctrl-a ctrl-d           # Detach from current session
ctrl-a ctrl-c           # Create new window in session
ctrl-a ctrl-n           # Go to next window
ctrl-a ctrl-p           # Go to previous window
ctrl-a ctrl-a           # Go to previously selected window
ctrl-a ?                # Display help</code></pre>
<pre class="bash"><code>screen -S session ID
# press return
# and enter your command or script
time populations ...

# you can close the terminal window, go home and re-open the terminal window ...

screen -ls
screen -r session ID</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
